tests/plugins/links/test_urlize.py

```diff
@@ -0,0 +1,231 @@
+    ),
+    (
+        '10.1.1.1:8000',
+        EXPECTED_PARAGRAPH_TEMPLATE % ('http://10.1.1.1:8000', '10.1.1.1:8000')
+    ),
+
+    # Test trailing path specification.
+    (
+        'http://example.com/',
+        EXPECTED_PARAGRAPH_TEMPLATE % ('http://example.com/', 'http://example.com/')
+    ),
+    (
+        'http://example.com/my/path',
+        EXPECTED_PARAGRAPH_TEMPLATE % ('http://example.com/my/path', 'http://example.com/my/path')
+    ),
+    (
+        'http://example.com/my/path?param1=value1&param2=value2',
+        EXPECTED_PARAGRAPH_TEMPLATE % ('http://example.com/my/path?param1=value1&amp;param2=value2', 'http://example.com/my/path?param1=value1&amp;param2=value2')
+    ),
+
+    # Link positioned somewhere within the text, but around whitespace boundary.
+    (
+        'This is link myhost.example.com',
+        "<p>This is link " + EXPECTED_LINK_TEMPLATE % ('http://myhost.example.com', 'myhost.example.com') + "</p>"
+    ),
+    (
+        'myhost.example.com is the link',
+        "<p>" + EXPECTED_LINK_TEMPLATE % ('http://myhost.example.com', 'myhost.example.com') + " is the link</p>"
+    ),
+    (
+        'I have best myhost.example.com link ever',
+        "<p>I have best " + EXPECTED_LINK_TEMPLATE % ('http://myhost.example.com', 'myhost.example.com') + " link ever</p>"
+    ),
+    (
+        'I have best\nmyhost.example.com link ever',
+        "<p>I have best\n" + EXPECTED_LINK_TEMPLATE % ('http://myhost.example.com', 'myhost.example.com') + " link ever</p>"
+    ),
+]
+
+
+FIXTURE_NEGATIVE_MATCHES = [
+    # Incomplete FQDNs.
+    (
+        'example.',
+        '<p>example.</p>'
+    ),
+    (
+        '.example .com',
+        '<p>.example .com</p>'
+    ),
+
+    # localhost as part of another word.
+    (
+        'localhosts',
+        '<p>localhosts</p>'
+    ),
+
+    # Invalid FQDNs.
+    (
+        'example-.com',
+        '<p>example-.com</p>'
+    ),
+    (
+        '-example.com',
+        '<p>-example.com</p>'
+    ),
+    (
+        'my.-example.com',
+        '<p>my.-example.com</p>'
+    ),
+
+    # Invalid IPv6 patterns.
+    (
+        '1:2:3:4:5:6:7:8:a',  # Use :a, because using a number would match as optional port
+        '<p>1:2:3:4:5:6:7:8:a</p>',
+    ),
+    (
+        '1::2::3',
+        '<p>1::2::3</p>',
+    ),
+    (
+        '::::1',
+        '<p>::::1</p>',
+    ),
+    (
+        '1::::',
+        '<p>1::::</p>',
+    ),
+
+    # Invalid IPv4 patterns.
+    (
+        '1.2.3.4.5',
+        '<p>1.2.3.4.5</p>',
+    ),
+
+]
+
+
+class TestUrlizeExtension:
+
+    def setup_method(self):
+        self.md = markdown.Markdown(extensions=[UrlizeExtension()])
+
+    @pytest.mark.parametrize("markdown_text, expected_output", FIXTURE_POSITIVE_MATCHES)
+    def test_positive_matches(self, markdown_text, expected_output):
+        assert self.md.convert(markdown_text) == expected_output
+
+    @pytest.mark.parametrize("markdown_text, expected_output", FIXTURE_NEGATIVE_MATCHES)
```

Comment:
Isn't this equivalent/comparable to `unittest.subTest`? I'm not sure how @benjaoming feels about different styles in the test suite, but I think choosing one consistent style is easier to maintain.

Reply:
Yup, I would agree that if `unittest` can do it, that's option 1.. then what `pytest` can offer would be second choice. But I think it's highly highly imaginative to think that one would want to change from `pytest` in a near future. Seems unlikely that a better test suite pops up or that `pytest` makes breaking API changes.

Reply:
I'll have a look at what unittest has to offer in terms of parametrisation. This stuff could be done in loop as well, but thought it was a nice thing to take advantage of functionality.

Reply:
Just had a look at it, that's a nice addition to unittest. I still find pytest a little bit superior, but better to keep everything consistent :)

Reply:
Ok, I take it back a bit. There's a [known missing support for `subTest`](https://github.com/pytest-dev/pytest/issues/1367) in `pytest`.

So, with pytest variant all possible things will be run even if some of them fail, with `pytest + subTest`, tests will stop after first failure.

So... Pick your poison, gentlemen :)

I could "worst case" convert everything to TestCase etc, but still use `pytest`'s parametrisation.

Reply:
@azaghal then I think it's totally fair to leave it as it is :) I don't want to impose some really rigid way of cutting out obvious benefits of using `pytest` -- after all, it's there to improve testing.

Reply:
Sheesh, ok, another update [from official pytest docs](https://docs.pytest.org/en/latest/unittest.html#pytest-features-in-unittest-testcase-subclasses) - the parametrize thing is not supported.

So... Might need to stick to `pytest`-style unit test unless you're fine with test stopping on first failure?

Again a matter of choose your poison thing, just let me know :)

Reply:
I'll be leaving the tests as `pytest`-tailored.